File: packages/feature-amove/src/main/legacy-migration.ts
import type { AppSettings, ShortcutBinding, ShortcutModifier, WindowActionId } from "../shared/contracts";
import { defaultSettings } from "../shared/defaults";
import { physicalCodeForCarbon } from "../shared/shortcuts";

export interface MigrationResult {
  settings: AppSettings;
  warning?: string;
}

const actionAliases: Record<string, WindowActionId> = {
  moveDisplayLeft: "moveDisplayLeft",
  previousDisplay: "moveDisplayLeft",
  moveDisplayRight: "moveDisplayRight",
  nextDisplay: "moveDisplayRight",
  moveDisplayUp: "moveDisplayUp",
  moveDisplayDown: "moveDisplayDown",
  toggleShelf: "toggleShelf"
};

export function migrateMacUserDefaults(values: Record<string, unknown>): MigrationResult {
  const settings = defaultSettings("darwin");
  const warnings: string[] = [];
  const shortcutSource = values["Amove.shortcuts.v2"] ?? values["ShiftShelf.shortcuts.v2"];

  if (shortcutSource !== undefined) {
    try {
      const decoded = decodeShortcutDocument(shortcutSource);
      const imported: Partial<Record<WindowActionId, ShortcutBinding>> = {};
      for (const [rawAction, rawBinding] of Object.entries(decoded)) {
        const action = actionAliases[rawAction];
        if (!action) {
          warnings.push(`Ignored unknown shortcut action “${rawAction}”.`);
          continue;
        }
        imported[action] = decodeLegacyBinding(rawBinding);
      }
      settings.shortcutsByPlatform.darwin = { ...settings.shortcutsByPlatform.darwin, ...imported };
    } catch (error) {
      warnings.push(`Shortcuts could not be imported: ${errorMessage(error)}`);
    }
  }

  try {
    settings.presenceMode = migratePresence(values);
  } catch (error) {
    warnings.push(`App presence could not be imported: ${errorMessage(error)}`);
  }

  settings.migrations.macUserDefaultsV1 = warnings.length === 0 ? "completed" : "completed-with-warnings";
  return warnings.length === 0 ? { settings } : { settings, warning: warnings.join(" ") };
}

function decodeShortcutDocument(value: unknown): Record<string, unknown> {
  if (isPlainRecord(value)) return value;
  let text: string;
  if (typeof value === "string") {
    const trimmed = value.trim();
    text = trimmed.startsWith("{") ? trimmed : Buffer.from(trimmed, "base64").toString("utf8");
  } else if (Buffer.isBuffer(value) || value instanceof Uint8Array) {
    text = Buffer.from(value).toString("utf8");
  } else {
    throw new Error("unsupported preference value");
  }
  const decoded: unknown = JSON.parse(text);
  if (!isPlainRecord(decoded)) throw new Error("shortcut document is not an object");
  return decoded;
}

function decodeLegacyBinding(value: unknown): ShortcutBinding {
  if (!isPlainRecord(value) || !Number.isInteger(value.keyCode)) throw new Error("shortcut entry is malformed");
  const keyCode = value.keyCode as number;
  const rawModifiers = isPlainRecord(value.modifiers) ? value.modifiers.rawValue : value.modifiers;
  if (!Number.isInteger(rawModifiers)) throw new Error("shortcut modifiers are malformed");
  const modifiers = decodeModifierMask(rawModifiers as number);
  const physicalCode = physicalCodeForCarbon(keyCode);
  return {
    key: physicalCode ? { kind: "physical", code: physicalCode } : { kind: "darwinCarbon", keyCode },
    modifiers
  };
}

function decodeModifierMask(mask: number): ShortcutModifier[] {
  const modifiers: ShortcutModifier[] = [];
  if ((mask & (1 << 2)) !== 0) modifiers.push("control");
  if ((mask & (1 << 1)) !== 0) modifiers.push("alt");
  if ((mask & (1 << 3)) !== 0) modifiers.push("shift");
  if ((mask & 1) !== 0) modifiers.push("meta");
  return modifiers;
}

function migratePresence(values: Record<string, unknown>): "background" | "taskbar" {
  const currentMenu = optionalBoolean(values["Amove.showsMenuBar"]);
  const currentDock = optionalBoolean(values["Amove.showsDockIconWhenNoWindowOpen"]);
  if (currentMenu !== undefined || currentDock !== undefined) {
    return currentDock === true && currentMenu === false ? "taskbar" : "background";
  }

  const legacyMenu = optionalBoolean(values["ShiftShelf.showsMenuBar"]);
  const legacyDock = optionalBoolean(values["ShiftShelf.showsDockIconWhenNoWindowOpen"]);
  if (legacyMenu !== undefined || legacyDock !== undefined) {
    return legacyDock === true && legacyMenu === false ? "taskbar" : "background";
  }

  const hideFromDock = optionalBoolean(values["ShiftShelf.hideFromDock"]);
  return hideFromDock === false ? "taskbar" : "background";
}

function optionalBoolean(value: unknown): boolean | undefined {
  if (value === undefined) return undefined;
  if (typeof value !== "boolean") throw new Error("expected a Boolean preference");
  return value;
}

function isPlainRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value) && !Buffer.isBuffer(value);
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: native/moirasia-runtime/Sources/AmoveRuntime/LegacyMigration.swift
import Foundation
import MoirasiaProtocol

/// Reads the preferences written by the legacy standalone Amove/ShiftShelf apps
/// and maps them onto the suite settings shape (parity with the UI migration).
public enum AmoveLegacyMigration {
    public static let legacyKeys = [
        "Amove.shortcuts.v2", "ShiftShelf.shortcuts.v2",
        "Amove.showsMenuBar", "Amove.showsDockIconWhenNoWindowOpen",
        "ShiftShelf.showsMenuBar", "ShiftShelf.showsDockIconWhenNoWindowOpen",
        "ShiftShelf.hideFromDock"
    ]

    /// Reads legacy CFPreferences values from the `com.opense.Amove` domain.
    public static func readLegacyPreferences() -> [String: JSONValue] {
        let domain = "com.opense.Amove" as CFString
        var result: [String: JSONValue] = [:]
        for key in legacyKeys {
            guard let value = CFPreferencesCopyAppValue(key as CFString, domain) else { continue }
            if let converted = convert(value) { result[key] = converted }
        }
        return result
    }

    public static func migrateMacUserDefaults(_ values: [String: JSONValue]) -> (settings: AmoveSettings, warnings: [String]) {
        var settings = AmoveSettings.darwinDefaults()
        var warnings: [String] = []
        let shortcutSource = values["Amove.shortcuts.v2"] ?? values["ShiftShelf.shortcuts.v2"]

        if let shortcutSource {
            switch decodeShortcutDocument(shortcutSource) {
            case .success(let document):
                let imported = document.compactMapValues { entry -> AmoveShortcutBinding? in
                    decodeLegacyBinding(entry)
                }
                var bindings = settings.shortcutsByPlatform["darwin"] ?? [:]
                for (rawAction, binding) in imported {
                    guard let action = actionAliases[rawAction] else {
                        warnings.append("Ignored unknown shortcut action “\(rawAction)”.")
                        continue
                    }
                    bindings[action] = binding
                }
                settings.shortcutsByPlatform["darwin"] = bindings
            case .failure(let message):
                warnings.append("Shortcuts could not be imported: \(message.message).")
            }
        }

        switch migratePresence(values) {
        case .success(let mode):
            settings.presenceMode = mode
        case .failure:
            warnings.append("App presence could not be imported.")
        }

        settings.migrations["macUserDefaultsV1"] = warnings.isEmpty ? "completed" : "completed-with-warnings"
        return (settings, warnings)
    }

    // MARK: - Internals

    private static let actionAliases: [String: String] = [
        "moveDisplayLeft": "moveDisplayLeft", "previousDisplay": "moveDisplayLeft",
        "moveDisplayRight": "moveDisplayRight", "nextDisplay": "moveDisplayRight",
        "moveDisplayUp": "moveDisplayUp", "moveDisplayDown": "moveDisplayDown",
        "toggleShelf": "toggleShelf"
    ]

    private static func decodeShortcutDocument(_ value: JSONValue) -> Result<[String: JSONValue], ShortcutDocumentError> {
        switch value {
        case .object(let object):
            return .success(object)
        case .string(let text):
            let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.hasPrefix("{") { return decodeJSON(trimmed) }
            guard let decoded = Data(base64Encoded: trimmed) else { return .failure(.init(message: "unsupported preference value")) }
            return decodeJSON(String(decoding: decoded, as: UTF8.self))
        default:
            return .failure(.init(message: "unsupported preference value"))
        }
    }

    private static func decodeJSON(_ text: String) -> Result<[String: JSONValue], ShortcutDocumentError> {
        guard let data = text.data(using: .utf8), let value = try? JSONDecoder().decode(JSONValue.self, from: data), case .object(let object) = value else {
            return .failure(.init(message: "shortcut document is not an object"))
        }
        return .success(object)
    }

    private struct ShortcutDocumentError: Error {
        let message: String
    }

    private static func decodeLegacyBinding(_ value: JSONValue) -> AmoveShortcutBinding? {
        guard case .object(let object) = value else { return nil }
        guard let keyCode = object["keyCode"]?.numberValue.map(Int.init) else { return nil }
        var rawMask: Int?
        if case .number(let mask) = object["modifiers"] { rawMask = Int(mask) }
        else if case .object(let wrapper) = object["modifiers"], let mask = wrapper["rawValue"]?.numberValue { rawMask = Int(mask) }
        guard let rawMask else { return nil }
        let modifiers = decodeModifierMask(rawMask)
        let physical = AmovePhysicalKeyCodes.physicalCodeForCarbon(UInt32(clamping: keyCode))
        let key = physical.map { AmoveShortcutKey(code: $0) } ?? AmoveShortcutKey(kind: "darwinCarbon", keyCode: keyCode)
        return AmoveShortcutBinding(key: key, modifiers: modifiers)
    }

    private static func decodeModifierMask(_ mask: Int) -> [String] {
        var modifiers: [String] = []
        if mask & (1 << 2) != 0 { modifiers.append("control") }
        if mask & (1 << 1) != 0 { modifiers.append("alt") }
        if mask & (1 << 3) != 0 { modifiers.append("shift") }
        if mask & 1 != 0 { modifiers.append("meta") }
        return modifiers
    }

    private static func migratePresence(_ values: [String: JSONValue]) -> Result<String, NoLegacyPresence> {
        let currentMenu = optionalBoolean(values["Amove.showsMenuBar"])
        let currentDock = optionalBoolean(values["Amove.showsDockIconWhenNoWindowOpen"])
        if currentMenu != nil || currentDock != nil {
            return .success(currentDock == true && currentMenu == false ? "taskbar" : "background")
        }
        let legacyMenu = optionalBoolean(values["ShiftShelf.showsMenuBar"])
        let legacyDock = optionalBoolean(values["ShiftShelf.showsDockIconWhenNoWindowOpen"])
        if legacyMenu != nil || legacyDock != nil {
            return .success(legacyDock == true && legacyMenu == false ? "taskbar" : "background")
        }
        let hideFromDock = optionalBoolean(values["ShiftShelf.hideFromDock"])
        return .success(hideFromDock == false ? "taskbar" : "background")
    }

    private struct NoLegacyPresence: Error {}

    private static func optionalBoolean(_ value: JSONValue?) -> Bool? {
        guard case .bool(let bool) = value else { return nil }
        return bool
    }

    private static func convert(_ value: CFTypeRef) -> JSONValue? {
        let typeID = CFGetTypeID(value)
        if typeID == CFBooleanGetTypeID() {
            return .bool(CFBooleanGetValue(value as! CFBoolean))
        }
        if typeID == CFNumberGetTypeID() {
            var number: Double = 0
            CFNumberGetValue(value as! CFNumber, .doubleType, &number)
            return .number(number)
        }
        if typeID == CFDataGetTypeID() {
            let data = value as! CFData
            let bytes = CFDataGetBytePtr(data)
            let length = CFDataGetLength(data)
            let buffer = length > 0 && bytes != nil ? Data(bytes: bytes!, count: length) : Data()
            return .string(buffer.base64EncodedString())
        }
        if typeID == CFStringGetTypeID() {
            return .string(value as! String)
        }
        return nil
    }
}